Use tempfile for creating a file in home directory for a IO test

chrispe opened the pull request:
- [X] closes #40091
- [X] tests added / passed
- [X] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [X] whatsnew entry

chrispe committed:
Add usage of tempfile

chrispe committed:
Use 'with' for tmp and use read mode for handle

jreback commented on pandas/tests/io/test_common.py:
```diff
@@ -119,10 +121,11 @@ def test_infer_compression_from_path(self, extension, expected, path_type):
-        with icom.get_handle(filename, "w") as handles:
-            assert os.path.isabs(handles.handle.name)
-            assert os.path.expanduser(filename) == handles.handle.name
+        with tempfile.NamedTemporaryFile(dir=os.path.expanduser("~/")) as tmp:
```
use tm.ensure_clean

chrispe replied:
I agree, but current implementation of `tm.ensure_clean` doesn't support custom directories. Should we then first extend that function?

jreback replied:
ok i guess this is fine. can you use pathlib.Path instead

chrispe committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into fix-issue-40091

chrispe committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into fix-issue-40091

chrispe committed:
Use pathlib.Path instead of os.path

chrispe committed:
Use hardcoded path for filename

Since we want to test for hardcoded case of "~", we should not use the Path.home()

chrispe committed:
Cast filename with path_type

chrispe committed:
1st attempt to solve windows permissions error

jreback commented:
lgtm ping on greenish (there is a failing numpy_dev build ATM)

chrispe committed:
Create temp dir to avoid race condition on windows

chrispe committed:
Reverted back to assert with os.path

chrispe committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into fix-issue-40091

jreback commented:
thanks @chrispe very nice!
